ulog.cpp

```diff
@@ -0,0 +1,316 @@
+
+    /* Check if we are interested in this msg_id */
+    if (msg_id != MAVLINK_MSG_ID_COMMAND_ACK && msg_id != MAVLINK_MSG_ID_LOGGING_DATA_ACKED
+        && msg_id != MAVLINK_MSG_ID_LOGGING_DATA) {
+        return buffer->len;
+    }
+
+    /*
+     * As MAVLINK_MSG_ID_LOGGING_DATA_ACKED and MAVLINK_MSG_ID_LOGGING_DATA
+     * are MAVLink 2 only messages lets untrim all messages that match
+     * with the interested msg_ids.
+     */
+    uint8_t data[MAVLINK_MAX_PACKET_LEN * 4];
+    struct buffer msg_buffer { 0, data };
+    uint8_t *payload;
+
+    untrim_msg(buffer, &msg_buffer);
+    buffer = &msg_buffer;
+
+    if (mavlink2)
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink2_header);
+    else
+        payload = buffer->data + sizeof(struct mavlink_router_mavlink1_header);
+
+    /* Handle messages */
+    switch (msg_id) {
+    case MAVLINK_MSG_ID_COMMAND_ACK: {
+        mavlink_command_ack_t *cmd = (mavlink_command_ack_t *)payload;
+
+        if (!_timeout || cmd->command != MAV_CMD_LOGGING_START)
+            break;
+
+        if (cmd->result == MAV_RESULT_ACCEPTED) {
+            _timeout_manager->del_timeout(_timeout);
+            _timeout = NULL;
+        } else
+            log_error("MAV_CMD_LOGGING_START result(%u) is different than accepted", cmd->result);
+        break;
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA_ACKED: {
+        mavlink_logging_data_acked_t *ulog_data_acked = (mavlink_logging_data_acked_t *)payload;
+        mavlink_message_t msg;
+        mavlink_logging_ack_t ack;
+
+        ack.sequence = ulog_data_acked->sequence;
+        ack.target_component = MAV_COMP_ID_ALL;
+        ack.target_system = _target_system_id;
+        mavlink_msg_logging_ack_encode(SYSTEM_ID, MAV_COMP_ID_ALL, &msg, &ack);
+        _send_msg(&msg, _target_system_id);
+    }
+    case MAVLINK_MSG_ID_LOGGING_DATA: {
+        mavlink_logging_data_t *ulog_data = (mavlink_logging_data_t *)payload;
+        _logging_data_process(ulog_data);
+        break;
+    }
+    }
+
+    return buffer->len;
+}
+
+bool ULog::_logging_seq(uint16_t seq, bool *drop)
+{
+    /* Check for message drops */
+    if (_expected_seq != seq) {
+        *drop = true;
+
+        /*
+         * Handle case were Flight Stack did not received a ACK and sent the
+         * same message again.
+         */
+        if (_expected_seq - 1 == seq)
+            return false;
+
+        _expected_seq = seq;
+    }
+
+    _expected_seq++;
+    return true;
+}
+
+void ULog::_logging_data_process(mavlink_logging_data_t *msg)
+{
+    bool drops = false;
+
+    if (!_logging_seq(msg->sequence, &drops))
+        return;
+
+    /* Waiting for ULog header? */
+    if (_waiting_header) {
+        const uint8_t magic[] = ULOG_MAGIC;
+
+        if (memcmp(magic, msg->data, sizeof(magic))) {
+            log_error("Invalid ULog Magic number");
+            return;
+        }
+
+        _waiting_header = false;
+        fwrite(msg->data, 1, ULOG_HEADER_SIZE, _file);
```

Comment:
you should only ever write to the final place if you are certain you received everything.  Or you need a mechanism to undo what you've done before. Imagine the situation in which you received a partial write, you wrote the header but you can't complete the message. You will have a wrong log.